gpbft/participant_test.go

```diff
@@ -0,0 +1,736 @@
+				nonNilJustification := &gpbft.Justification{}
+				return []*gpbft.GMessage{
+					{
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.PREPARE_PHASE,
+						},
+						Signature:     signature,
+						Justification: nonNilJustification,
+					},
+					{
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.QUALITY_PHASE,
+							Value:    subject.canonicalChain,
+						},
+						Signature:     signature,
+						Justification: nonNilJustification,
+					},
+					{
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.COMMIT_PHASE,
+						},
+						Signature:     signature,
+						Justification: nonNilJustification,
+					},
+				}
+			},
+			wantErr: "has unexpected justification",
+		},
+		{
+			name: "nil Justification when needed is error",
+			msgs: func(subject *participantTestSubject) []*gpbft.GMessage {
+				subject.mockValidSignature(somePowerEntry.PubKey, signature)
+				return []*gpbft.GMessage{
+					{
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.DECIDE_PHASE,
+							Value:    subject.canonicalChain,
+						},
+						Signature: signature,
+					},
+					{
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.COMMIT_PHASE,
+							Value:    subject.canonicalChain,
+						},
+						Signature: signature,
+					},
+				}
+			},
+			wantErr: "has no justification",
+		},
+		{
+			name: "CONVERGE with nil Justification when needed is error",
+			msgs: func(subject *participantTestSubject) []*gpbft.GMessage {
+				ticket := gpbft.Ticket("fishcake")
+				subject.mockValidTicket(somePowerEntry.PubKey, ticket)
+				subject.mockValidSignature(somePowerEntry.PubKey, signature)
+				return []*gpbft.GMessage{
+					{
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.CONVERGE_PHASE,
+							Round:    4,
+							Value:    subject.canonicalChain,
+						},
+						Ticket:    ticket,
+						Signature: signature,
+					},
+				}
+			},
+			wantErr: "has no justification",
+		},
+		{
+			name: "justification and vote instance mismatch is error",
+			msgs: func(subject *participantTestSubject) []*gpbft.GMessage {
+				subject.mockValidSignature(somePowerEntry.PubKey, signature)
+				return []*gpbft.GMessage{
+					{
+						Sender: somePowerEntry.ID,
+						Vote: gpbft.Payload{
+							Instance: initialInstanceNumber,
+							Step:     gpbft.COMMIT_PHASE,
+							Value:    subject.canonicalChain,
+						},
+						Signature: signature,
+						Justification: &gpbft.Justification{
+							Vote: gpbft.Payload{
+								Instance: initialInstanceNumber + 3,
```

Comment:
It's worth testing all possible mismatches in vote/justification here. This is a high stakes piece of code.

Reply:
Thank you for pointing that out. 
I have (non-exhaustively) added a few more cases here. Full permutation of possible cases would need a bit more templating to reduce duplicate code. I have taken note to address this in follow up PRs as this one is verbose enough already.